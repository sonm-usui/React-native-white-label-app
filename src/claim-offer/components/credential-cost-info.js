// @flow
import React from 'react'
import { View, StyleSheet, FlatList } from 'react-native'
import { scale } from 'react-native-size-matters'
import { BigNumber } from 'bignumber.js'

import {
  cardBorder,
  yellowSea,
  whiteSolid,
  isBiggerThanMediumDevice, colors,
} from '../../common/styles'
import { ModalButtons } from '../../components/buttons/modal-buttons'
import type { TokenFeesData } from '../type-claim-offer'
import { CustomText } from '../../components/'

type CredentialCostInfoProps = {
  feesData: TokenFeesData,
  payTokenValue: string,
  backgroundColor: string,
  onConfirmAndPay: () => void,
  onCancel: () => void,
  secondColorBackground: string,
}

const CredentialCostInfo = (props: CredentialCostInfoProps) => {
  const {
    feesData,
    payTokenValue,
    backgroundColor,
    onCancel,
    onConfirmAndPay,
    secondColorBackground,
  } = props
  const credentialCost = new BigNumber(payTokenValue || 0).toFixed(8)
  const costsData = [
    {
      label: 'Transaction Fee',
      key: 'fees',
      value: feesData.fees,
    },
    {
      label: 'Credential Cost',
      key: 'payTokenValue',
      value: credentialCost,
    },
    { label: 'Total', key: 'total', value: feesData.total || 0 },
  ]

  const renderCostCell = ({ item, index }) => {
    const { totalLabel, totalValue, labelText, valueText } = styles
    let totalLabelStyles = [totalLabel]
    let totalValueStyles = [totalValue]
    if (index === 2 && item.value && item.value.length >= 12) {
      totalLabelStyles.push({ fontSize: scale(17) })
      totalValueStyles.push({ fontSize: scale(19) })
    }

    return (
      <View style={styles.cell} key={item.label}>
        <CustomText
          bg={false}
          style={[labelText, index === 2 && totalLabelStyles]}
        >
          {item && item.label}
        </CustomText>
        <CustomText
          bg={false}
          formatNumber
          style={[valueText, index === 2 && totalValueStyles]}
        >
          {item && item.value}
        </CustomText>
      </View>
    )
  }

  const borderSeparator = ({ leadingItem }) => {
    const { key } = leadingItem
    return (
      <View
        style={[
          styles.borderSeparator,
          key === 'payTokenValue' && styles.borderTotalSeparator,
        ]}
      />
    )
  }

  const keyExtractor = (item) => `${item.label}`

  return (
    <View style={styles.costContainer}>
      <CustomText bg={false} style={[styles.noteMessage]}>
        Please confirm. Once tokens are transferred, it cannot be undone.
      </CustomText>
      {/* NOTE: Probably don't need to scroll here, should be able to make UI fit on screen */}
      <FlatList
        data={costsData}
        style={styles.costTable}
        keyExtractor={keyExtractor}
        ItemSeparatorComponent={borderSeparator}
        renderItem={renderCostCell}
      />
      {/* <View style={styles.buttonsContainer}> */}
      <ModalButtons
        onIgnore={onCancel}
        onPress={onConfirmAndPay}
        disableAccept={false}
        colorBackground={backgroundColor}
        secondColorBackground={secondColorBackground}
        denyButtonText={'Cancel'}
        acceptBtnText={'Confirm and Pay'}
      />
      {/* </View> */}
    </View>
  )
}

export default CredentialCostInfo

const textStyle = {
  color: colors.gray1,
  lineHeight: 20,
}

const styles = StyleSheet.create({
  costContainer: { flex: 1, backgroundColor: cardBorder },
  costTable: {
    height: 'auto',
    // flex: 1,
  },
  noteMessage: {
    color: colors.gray2,
    textAlign: 'center',
    fontSize: isBiggerThanMediumDevice ? scale(15) : scale(12),
    paddingTop: isBiggerThanMediumDevice ? 15 : 5,
    paddingBottom: isBiggerThanMediumDevice ? 10 : 5,
    flex: 1,
  },
  cell: {
    padding: 15,
    flexDirection: 'row',
    justifyContent: 'space-between',
    backgroundColor: whiteSolid,
  },
  valueText: {
    ...textStyle,
    fontSize: scale(17),
    color: yellowSea,
    textAlign: 'right',
  },
  labelText: {
    fontWeight: 'bold',
    fontSize: scale(14),
  },
  borderSeparator: {
    backgroundColor: colors.gray4,
    width: '100%',
    height: 1,
  },
  borderTotalSeparator: {
    backgroundColor: yellowSea,
    width: '100%',
    height: 2,
  },
  totalLabel: {
    fontSize: scale(19),
    lineHeight: 22,
    textTransform: 'uppercase',
  },
  totalValue: {
    fontSize: scale(22),
    lineHeight: 22,
    textTransform: 'uppercase',
    fontWeight: '700',
  },
  buttonsContainer: {
    flex: 1,
  },
})
