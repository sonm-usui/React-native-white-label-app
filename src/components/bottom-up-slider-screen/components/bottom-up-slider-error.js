// @flow

import React from 'react'
import { StyleSheet } from 'react-native'
import LottieView from 'lottie-react-native'

import { CustomView } from '../../layout/custom-view'
import { BottomUpSliderText } from './bottom-up-slider-screen-text'

export function BottomUpSliderError(props: {
  afterErrorShown?: () => void,
  errorText: string,
  textStyles?: Array<Object | number>,
  errorIconStyles?: Array<Object | number>,
  containerStyles?: Array<Object | number>,
}) {
  const { textStyles = [], errorIconStyles = [], containerStyles = [] } = props
  return (
    <CustomView
      bg="tertiary"
      center
      style={[styles.errorContainer, ...containerStyles]}
    >
      <CustomView>
        <LottieView
          source={require('../../../images/red-cross-lottie.json')}
          autoPlay
          loop={false}
          style={[styles.feedbackIcon, ...errorIconStyles]}
          onAnimationFinish={props.afterErrorShown}
        />
      </CustomView>
      <BottomUpSliderText size="h4" bold={false} center style={[...textStyles]}>
        {props.errorText}
      </BottomUpSliderText>
    </CustomView>
  )
}

const styles = StyleSheet.create({
  errorContainer: {},
  feedbackIcon: {
    width: 50,
    height: 50,
  },
})
