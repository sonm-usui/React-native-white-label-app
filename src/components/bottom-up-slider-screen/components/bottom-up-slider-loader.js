// @flow

import React from 'react'
import { StyleSheet } from 'react-native'

import { CustomView } from '../../layout/custom-view'
import Loader from '../../loader/loader'
import type { Styles } from '../../../common/type-common'

export function BottomUpSliderLoader({
  message = 'Sending...',
  style,
}: {
  message?: string,
  style?: Styles,
}) {
  return (
    <CustomView bg="tertiary" center style={[styles.loaderContainer, style]}>
      <Loader type="dark" showMessage={true} message={message} />
    </CustomView>
  )
}

const styles = StyleSheet.create({
  loaderContainer: {
    minHeight: '20%',
  },
})
