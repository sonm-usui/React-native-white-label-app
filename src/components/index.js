// @flow

export { Avatar } from './avatar/avatar'
export { Container, CustomView, headerStyles } from './layout'
export { default as Icon } from './icon'
export { default as CustomText, styles as TextStyles } from './text'
export { default as CustomButton } from './button'
export { default as QRScanner } from './qr-scanner/qr-scanner'
export { default as TouchId } from './touch-id/touch-id'
export { default as PinCodeBox } from './pin-code/pin-code-box'
export { default as ImageColorPicker } from './image-color-picker/image-color-picker'
export { default as ConnectionTheme } from './image-color-picker/connection-theme'
export { default as AvatarsPair } from './AvatarsPair'
export { default as Separator } from './separator'
export { default as FooterActions } from './footer-actions/footer-actions'
export { default as CustomListProofRequest } from './custom-list/custom-list-proof-request'
export { default as CustomList } from './custom-list/custom-list'
export { default as CustomDate } from './custom-date/custom-date'
export { default as UserAvatar } from './user-avatar/user-avatar'
export { default as Keyboard } from './keyboard/keyboard'
export { default as CustomHeader } from './custom-header/custom-header'
export { default as Loader } from './loader/loader'
export { Header } from './header/common-header/header'
export { HeaderWithMenu } from './header/header-with-menu/header-with-menu'
export { HeaderWithDeletion } from './header/header-with-deletion/header-with-deletion'
export { ConnectionCard } from '../my-connections/connection-card/connection-card'
export { CameraButton } from './camera-button/camera-button'
export { default as UnreadMessagesBadge } from './unread-messages-badge/unread-messages-badge'
