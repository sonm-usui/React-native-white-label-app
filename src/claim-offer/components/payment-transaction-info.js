// @flow
import * as React from 'react'
import { PureComponent } from 'react'
import { View, StyleSheet, ScrollView } from 'react-native'
import { scale } from 'react-native-size-matters'

import type {
  PaymentTransactionInfoProps,
  TokenFeesData,
} from '../type-claim-offer'

import CredentialCostInfo from './credential-cost-info'
import { ModalButtons } from '../../components/buttons/modal-buttons'
import { CustomText } from '../../components'
import { colors } from '../../common/styles/constant'
import Loader from '../../components/loader/loader'
import { Error } from '../../components/error/error'
import { Container } from '../../components'
import { Success } from '../../components/success/success'
import { CLAIM_REQUEST_STATUS } from '../type-claim-offer'
import { animateLayout } from '../../common/layout-animation'

const messages = {
  IN_PROGRESS: 'Getting network fees...',
  ERROR: 'There was a problem getting network fees.',
  ZERO_FEES: 'No network fees for this transaction. Transferring tokens...',
  INSUFFICIENT_BALANCE:
    'You do not have enough tokens to purchase this credential.',
  INSUFFICIENT_BALANCE_WITH_DATA: (
    feesData: TokenFeesData & { credentialPrice: string }
  ) =>
    `You do not have enough tokens to purchase this credential. You need (credential price: ${feesData.credentialPrice} + fees: ${feesData.fees} = total: ${feesData.total}), while your current token balance is ${feesData.currentTokenBalance}. Please get in touch with Sovrin Foundation to get more tokens. Once you have enough tokens, open credential offer again and you should be able to accept.`,
  SENDING_PAID_CREDENTIAL_REQUEST: 'Transferring tokens...',
  SENDING_CREDENTIAL_REQUEST: 'Accepting offer...',
  SUCCESS:
    'Tokens transferred successfully. You should receive your credential shortly.',
  SUCCESS_SEND_CLAIM_REQUEST: 'Offer accepted successfully.',
  ERROR_PAID_CREDENTIAL_REQUEST: 'There was a problem transferring tokens.',
  ERROR_CREDENTIAL_REQUEST: 'There was a problem accepting this offer.',
}

class PaymentTransactionInfo extends PureComponent<
  PaymentTransactionInfoProps,
  void
> {
  render() {
    const {
      claimThemePrimary,
      claimThemeSecondary,
      onConfirmAndPay,
      onCancel,
      credentialPrice,
      txnFeesStatus,
      claimRequestStatus,
      onRetry,
      feesData = {
        fees: '0',
        total: '',
        currentTokenBalance: '',
      },
      onSuccess = noop,
    } = this.props

    let content = null
    let isError = false

    if (txnFeesStatus) {
      const credentialCostInfo = (
        <CredentialCostInfo
          feesData={feesData}
          payTokenValue={credentialPrice}
          backgroundColor={claimThemePrimary}
          onConfirmAndPay={onConfirmAndPay}
          onCancel={onCancel}
          secondColorBackground={claimThemeSecondary}
        />
      )
      const txnFeesComponent = {
        IN_PROGRESS: <Loader message={messages.IN_PROGRESS} />,
        ERROR: (
          <Error
            errorText={messages.ERROR}
            containerStyles={[styles.errorContainer]}
          />
        ),
        ZERO_FEES: (
          <Container center>
            <Text>{messages.ZERO_FEES}</Text>
          </Container>
        ),
        TRANSFER_EQUAL_TO_BALANCE: credentialCostInfo,
        TRANSFER_POSSIBLE_WITH_FEES: credentialCostInfo,
        TRANSFER_NOT_POSSIBLE_WITH_FEES: (
          <ScrollView style={styles.scrollViewStyles}>
            <Error
              errorText={messages.INSUFFICIENT_BALANCE_WITH_DATA({
                ...feesData,
                credentialPrice,
              })}
              textStyles={[styles.insufficientBalanceText]}
              containerStyles={[styles.errorContainer]}
            />
          </ScrollView>
        ),
      }

      content = txnFeesComponent[txnFeesStatus]
      isError = ['ERROR', 'TRANSFER_NOT_POSSIBLE_WITH_FEES'].includes(
        txnFeesStatus
      )
    }

    if (claimRequestStatus) {
      switch (claimRequestStatus) {
        case CLAIM_REQUEST_STATUS.SENDING_PAID_CREDENTIAL_REQUEST:
          content = (
            <Loader message={messages.SENDING_PAID_CREDENTIAL_REQUEST} />
          )
          break
        case CLAIM_REQUEST_STATUS.SENDING_CLAIM_REQUEST:
          content = <Loader message={messages.SENDING_CREDENTIAL_REQUEST} />
          break
        case CLAIM_REQUEST_STATUS.PAID_CREDENTIAL_REQUEST_FAIL:
          content = (
            <Error
              errorText={messages.ERROR_PAID_CREDENTIAL_REQUEST}
              containerStyles={[styles.errorContainer]}
            />
          )
          break
        case CLAIM_REQUEST_STATUS.SEND_CLAIM_REQUEST_FAIL:
        case CLAIM_REQUEST_STATUS.CLAIM_REQUEST_FAIL:
          content = (
            <Error
              errorText={messages.ERROR_CREDENTIAL_REQUEST}
              containerStyles={[styles.errorContainer]}
            />
          )
          break
        case CLAIM_REQUEST_STATUS.INSUFFICIENT_BALANCE:
          content = (
            <Error
              errorText={messages.INSUFFICIENT_BALANCE}
              containerStyles={[styles.errorContainer]}
            />
          )
          break
        case CLAIM_REQUEST_STATUS.PAID_CREDENTIAL_REQUEST_SUCCESS:
          content = (
            <Success
              successText={messages.SUCCESS}
              afterSuccessShown={onSuccess}
            />
          )
          break
        case CLAIM_REQUEST_STATUS.SEND_CLAIM_REQUEST_SUCCESS:
          content = (
            <Success
              successText={messages.SUCCESS_SEND_CLAIM_REQUEST}
              afterSuccessShown={onSuccess}
            />
          )
          break
        default:
          break
      }

      isError = [
        CLAIM_REQUEST_STATUS.CLAIM_REQUEST_FAIL,
        CLAIM_REQUEST_STATUS.SEND_CLAIM_REQUEST_FAIL,
        CLAIM_REQUEST_STATUS.INSUFFICIENT_BALANCE,
        CLAIM_REQUEST_STATUS.PAID_CREDENTIAL_REQUEST_FAIL,
      ].includes(claimRequestStatus)
    }

    if (content) {
      return (
        <View style={styles.container}>
          {content}
          {isError ? (
            <ModalButtons
              onPress={onRetry}
              onIgnore={onCancel}
              colorBackground={claimThemePrimary}
              secondColorBackground={claimThemeSecondary}
              denyButtonText={'Cancel'}
              acceptBtnText={'Retry'}
            />
          ) : null}
        </View>
      )
    }

    return null
  }

  getSnapshotBeforeUpdate() {
    animateLayout()
  }
}

function Text({ children }: { children: React.Node }) {
  return (
    <CustomText
      primary
      multiline
      transparentBg
      center
      style={styles.statusText}
    >
      {children}
    </CustomText>
  )
}

function noop() {}

export default PaymentTransactionInfo

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: '5%',
    paddingTop: 20,
    backgroundColor: colors.gray5,
  },
  statusText: {
    color: colors.gray3,
    textAlign: 'center',
    marginTop: 40,
    fontSize: scale(15),
    backgroundColor: colors.red,
  },
  errorContainer: {
    flex: 1,
  },
  insufficientBalanceText: { fontSize: scale(17) },
  scrollViewStyles: {
    flex: 1,
  },
})
