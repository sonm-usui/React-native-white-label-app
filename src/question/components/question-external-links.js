// @flow
import React, { PureComponent } from 'react'
import { StyleSheet, TouchableOpacity, Alert, Linking } from 'react-native'
import Snackbar from 'react-native-snackbar'

import { PrivacyPolicyTitle } from '../../common/privacyTNC-constants'
import type { ExternalLink } from '../type-question'
import type { CustomError } from '../../common/type-common'

import {
  ERROR_EXTERNAL_LINKS_NOT_ARRAY,
  ERROR_EXTERNAL_LINKS_NOT_PROPERLY_FORMATTED,
  ERROR_TOO_MANY_EXTERNAL_LINKS,
} from '../type-question'
import { CustomView, CustomText } from '../../components'
import {
  OFFSET_1X,
  alertCancel,
  venetianRed,
  color, colors,
} from '../../common/styles'
import { appName } from '../../external-imports'

export class QuestionExternalLinks extends PureComponent<QuestionExternalLinksProps> {
  render() {
    const { externalLinks = [] } = this.props

    const validationError = getExternalLinkValidity(externalLinks)
    if (validationError) {
      return (
        <CustomView style={[styles.validationErrorContainer]}>
          <CustomText bg={false} size="h6" style={[styles.validationErrorText]}>
            {`Invalid external link. Error: ${validationError.code}`}
          </CustomText>
        </CustomView>
      )
    }

    if (!hasExternalLinks(externalLinks)) {
      return null
    }

    // we need to take only maximum links that we are allowed to show
    const maxLinks = externalLinks.slice(0, MAX_EXTERNAL_LINKS_TO_SHOW)

    return (
      <CustomView>
        {maxLinks.map((link: ExternalLink, index: number) => (
          <QuestionExternalLink key={`${link.src}${index}`} link={link} />
        ))}
      </CustomView>
    )
  }
}

export class QuestionExternalLink extends PureComponent<QuestionExternalLinkProps> {
  render() {
    const { text, src } = this.props.link
    return (
      <CustomView style={[styles.externalLinkContainer]}>
        <TouchableOpacity onPress={this.onLinkClick}>
          <CustomText
            bg={false}
            size="h6"
            style={[styles.externalLinkText]}
            numberOfLines={2}
          >
            {text || src}
          </CustomText>
        </TouchableOpacity>
      </CustomView>
    )
  }

  onLinkClick = () => {
    Alert.alert(
      `You are about to leave ${appName}`,
      `The contents of this link are not controlled by ${appName} and are not covered by its ${PrivacyPolicyTitle} and End User License Agreement. Proceed?`,
      [alertCancel, { text: 'OK', onPress: this.openLink }]
    )
  }

  openLink = () => {
    Linking.openURL(this.props.link.src).catch(() => {
      Snackbar.show({
        text: 'Could not open this link. Link seems to be incorrect.',
        duration: Snackbar.LENGTH_LONG,
        backgroundColor: venetianRed,
        textColor: colors.white,
      })
    })
  }
}

export function getExternalLinkValidity(
  externalLinks: ExternalLink[]
): null | CustomError {
  if (!Array.isArray(externalLinks)) {
    return ERROR_EXTERNAL_LINKS_NOT_ARRAY
  }

  if (externalLinks.length > 1000) {
    return ERROR_TOO_MANY_EXTERNAL_LINKS
  }

  const everyLinkValid = externalLinks.every((link: ExternalLink) => {
    // if an element of external_links array is either null or false or empty string
    // then it is not a valid element
    const linkJSType = typeof link
    if (
      !link ||
      ['string', 'number', 'boolean', 'function', 'symbol'].includes(
        linkJSType
      ) ||
      Array.isArray(link)
    ) {
      return false
    }

    const { text, src } = link
    if (text) {
      // if we do get text, then it should be a string
      // with less than MAX_CHARACTERS_FOR_TEXT characters
      if (typeof text !== 'string' || text.length > MAX_CHARACTERS_FOR_TEXT) {
        return false
      }
    }

    // src should be defined and should be a string
    return typeof src === 'string' && src
  })
  if (!everyLinkValid) {
    return ERROR_EXTERNAL_LINKS_NOT_PROPERLY_FORMATTED
  }

  return null
}

function hasExternalLinks(externalLinks: Array<ExternalLink>): boolean {
  return Array.isArray(externalLinks) && externalLinks.length > 0
}

const styles = StyleSheet.create({
  externalLinkContainer: {
    marginVertical: 8,
  },
  externalLinkText: {
    color: 'blue',
  },
  validationErrorContainer: {
    marginVertical: OFFSET_1X,
  },
  validationErrorText: {
    color: color.bg.tertiary.font.seventh,
  },
})

export const MAX_EXTERNAL_LINKS_TO_SHOW = 20
const MAX_CHARACTERS_FOR_TEXT = 1000

export type QuestionExternalLinksProps = {
  externalLinks: Array<ExternalLink>,
}

export type QuestionExternalLinkProps = {
  link: ExternalLink,
}
