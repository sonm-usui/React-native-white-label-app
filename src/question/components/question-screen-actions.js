// @flow

import React, { useCallback } from 'react'

import type { QuestionResponse, QuestionStoreMessage } from '../type-question'

import { CustomView, CustomButton } from '../../components'
import { TEXT_SUBMIT, TEXT_TRY_AGAIN, TEXT_OK } from '../type-question'
import { getScreenStatus } from '../question-store'
import {
  questionActionButtonDefaultProps,
  questionStyles,
} from '../question-screen-style'

import { colors, isSmallWidthDevice } from '../../common/styles'
import { ModalButtons } from '../../components/buttons/modal-buttons'

export const QuestionActions = ({
  question,
  selectedResponse,
  onSelectResponseAndSubmit,
  onSubmit,
  isOffline,
}: QuestionActionProps) => {
  const testID = 'question-action'

  if (!question) {
    return null
  }
  const { valid_responses: responses } = question.payload

  let submitButtonTitle = TEXT_SUBMIT

  const { error, success, idle } = getScreenStatus(
    question ? question.status : undefined
  )
  if (error) {
    submitButtonTitle = TEXT_TRY_AGAIN
  }

  if (success) {
    submitButtonTitle = TEXT_OK
  }
  // if screen status is success or error, we need to enable submit button
  // if status is idle, then user needs to first select an answer
  // and only after that submit button is enabled
  const isSubmitDisabled = success || error ? false : !selectedResponse

  // If there are more than two responses, then we need to render
  // our own buttons (submit and cancel)
  const shouldRenderSubmitButton = idle ? responses.length > 2 : true

  // we need to render responses as buttons, so user can just press a button
  // to select the response for question
  // If a question has two valid responses
  // then those two responses would be rendered as buttons in screen
  // and user can tap on either button to send that particular response
  const responseButtons = (
    <QuestionResponseButtons
      responses={responses}
      onPress={onSelectResponseAndSubmit}
      disabled={isOffline}
    />
  )

  return (
    <CustomView style={questionStyles.actionButtonContainer}>
      {shouldRenderSubmitButton ? (
        <ModalButtons
          onPress={onSubmit}
          acceptBtnText={submitButtonTitle}
          disableAccept={isSubmitDisabled || isOffline}
          colorBackground={colors.main}
          bottomTestID={`${testID}-submit`}
        />
      ) : (
        responseButtons
      )}
    </CustomView>
  )
}

function QuestionResponseButtons(props: QuestionResponseButtonsProps) {
  const { responses, onPress, disabled } = props

  if (responses.length === 0 || responses.length > 2) {
    return null
  }

  // if we have only single response, then we need some more business logic
  // as per our requirement, see <QuestionResponseButton /> render method
  // for more logic
  const isSingleResponse = responses.length === 1

  return (
    <CustomView style={[questionStyles.actionWrapper]}>
      {responses.map((response: QuestionResponse, index: number) => (
        <QuestionResponseButton
          response={response}
          key={response.nonce}
          onPress={onPress}
          // Also, only first (or last in reversed array) response in array
          // is supposed to look green (primary button)
          // and all other responses needs to look secondary
          isPrimaryResponse={index === 0}
          isSingleResponse={isSingleResponse}
          testID={response.text}
          accessible={true}
          accessibilityLabel={response.text}
          disabled={disabled}
        />
      ))}
    </CustomView>
  )
}

const QuestionResponseButton = ({
  response,
  isPrimaryResponse,
  isSingleResponse,
  onPress,
  disabled = false,
}: QuestionResponseButtonProps) => {
  const onResponse = useCallback(() => {
    onPress(response)
  }, [])
  // We need to show only one line of text on action buttons
  // Normally, we should have used react-native's numberOfLines prop
  // but since we are using an external library for Buttons
  // and this version is bit older, in newer versions we can specify
  // numberOfLines prop for Button component
  // So, for now just using substring on response text
  // Also, if there is only a single response
  // then we need to show half text limit on button, because in single response
  // buttons are placed side by side and has less width
  // TODO: Upgrade react-native-elements and use numberOfLines prop
  const buttonText = isSingleResponse
    ? ellipsis(response.text, Math.floor(getResponseButtonsTextLimit() / 2))
    : ellipsis(response.text, getResponseButtonsTextLimit())

  return (
    <CustomView
      style={[
        questionStyles.actionButtonContainer,
        isPrimaryResponse
          ? questionStyles.submitButton
          : questionStyles.cancelButton,
      ]}
    >
      <CustomButton
        disabled={disabled}
        {...questionActionButtonDefaultProps}
        main={isPrimaryResponse}
        twelfth={!isPrimaryResponse}
        style={[
          questionStyles.actionButton,
          isPrimaryResponse && questionStyles.submitButton,
        ]}
        title={buttonText}
        testID={response.text}
        accessible={true}
        accessibilityLabel={response.text}
        onPress={onResponse}
        titleStyle={isPrimaryResponse ? {} : questionStyles.cancelBtnColor}
      />
    </CustomView>
  )
}

function ellipsis(text: string, maxLength: number) {
  const stringLargerThanMaxLength = text.length > maxLength
  const trimmedText = text.substring(
    0,
    stringLargerThanMaxLength ? maxLength - 3 : maxLength
  )
  return stringLargerThanMaxLength ? trimmedText + '...' : trimmedText
}

function getResponseButtonsTextLimit() {
  const responseTextLimit = 40
  return isSmallWidthDevice ? responseTextLimit - 5 : responseTextLimit
}

export type QuestionActionProps = {
  selectedResponse: ?QuestionResponse,
  onSubmit: () => void,
  onSelectResponseAndSubmit: (response: QuestionResponse) => void,
  question?: QuestionStoreMessage,
  isOffline: boolean,
}

type QuestionResponseButtonsProps = {
  responses: Array<QuestionResponse>,
  onPress: (response: QuestionResponse) => void,
  disabled: boolean,
}

type QuestionResponseButtonProps = {
  response: QuestionResponse,
  onPress: (response: QuestionResponse) => void,
  isPrimaryResponse: boolean,
  isSingleResponse: boolean,
  disabled: boolean,
}
