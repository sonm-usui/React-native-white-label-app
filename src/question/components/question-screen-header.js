// @flow
import React, { PureComponent } from 'react'

import { Container, CustomView } from '../../components'
import { questionStyles } from '../question-screen-style'

export class QuestionScreenHeader extends PureComponent<
  QuestionScreenHeaderProps,
  void
> {
  render() {
    return (
      <Container row style={[questionStyles.headerContainer]}>
        <ViewCloser {...this.props} />
        {!this.props.hideHandlebar && <HeaderHandlebar />}
        <ViewCloser {...this.props} />
      </Container>
    )
  }
}

// this component is used to close screen, when user taps on header
export const ViewCloser = (props: QuestionScreenHeaderProps) => (
  <Container onPress={props.onCancel} />
)

export const HeaderHandlebar = () => (
  <CustomView style={[questionStyles.headerHandleContainer]}>
    <CustomView style={[questionStyles.headerHandlebar]} />
  </CustomView>
)

type QuestionScreenHeaderProps = {
  onCancel: () => void,
  hideHandlebar?: boolean,
}
