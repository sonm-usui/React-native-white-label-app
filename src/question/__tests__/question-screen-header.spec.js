// @flow

import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'

import {
  QuestionScreenHeader,
  ViewCloser,
} from '../components/question-screen-header'
import { Container } from '../../components'

describe('<QuestionScreenHeader />', () => {
  function getProps() {
    return {
      onCancel: jest.fn(),
    }
  }

  it('should match snapshot', () => {
    const props = getProps()
    const tree = renderer.create(<QuestionScreenHeader {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('click on header should trigger onCancel on header', () => {
    const props = getProps()
    const component = renderer.create(<QuestionScreenHeader {...props} />)
    const componentViewCloser = component.root.findAllByType(ViewCloser)
    componentViewCloser.map(viewCloser => {
      const container = viewCloser.findByType(Container)
      container.props.onPress()
    })
    expect(props.onCancel).toHaveBeenCalledTimes(componentViewCloser.length)
  })
})
