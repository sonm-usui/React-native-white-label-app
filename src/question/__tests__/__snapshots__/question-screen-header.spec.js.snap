// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionScreenHeader /> should match snapshot 1`] = `
<View
  row={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "justifyContent": "flex-end",
        "marginBottom": 8,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#F2F2F2",
          "borderRadius": 6,
          "height": 6,
          "width": 51,
        }
      }
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 20,
        "top": 20,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
